core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,396 @@
+    {
+    case EH_AES_GCM_128:
+    case EH_SM4:
+        return 16;
+    case EH_AES_GCM_192:
+        return 24;
+    case EH_AES_GCM_256:
+        return 32;
+    default:
+        return UINT32_MAX;
+    }
+    return UINT32_MAX;
+}
+
+/**
+ * @brief generate aes_gcm key with openssl api
+ * @param cmk_blob storage key information
+ * @param SIZE_OF_KEYBLOB_T the size of cmk_blob
+ * @param req_blob_size the gcm data size
+ * @param keyspec key type, refer to [ehsm_keyspec_t]
+ * @return sgx_status_t
+ */
+sgx_status_t ehsm_create_aes_key(uint8_t *cmk_blob, uint32_t SIZE_OF_KEYBLOB_T,
+                                 uint32_t *req_blob_size, ehsm_keyspec_t keyspec)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (keyspec != EH_AES_GCM_128 &&
+        keyspec != EH_AES_GCM_192 &&
+        keyspec != EH_AES_GCM_256)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint32_t keysize = ehsm_get_symmetric_key_size(keyspec);
+    if (keysize == UINT32_MAX)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    uint32_t real_blob_len = ehsm_calc_keyblob_len(0, keysize);
+
+    if (real_blob_len == UINT32_MAX)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    if (req_blob_size != NULL)
+    {
+        *req_blob_size = real_blob_len;
+        return SGX_SUCCESS;
+    }
+    if (cmk_blob == NULL || SIZE_OF_KEYBLOB_T != real_blob_len)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint8_t *tmp = (uint8_t *)malloc(keysize);
+    if (tmp == NULL)
+    {
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+    ret = sgx_read_rand(tmp, keysize);
+    if (ret != SGX_SUCCESS)
+    {
+        free(tmp);
+        return ret;
+    }
+    ret = ehsm_create_keyblob(tmp,
+                              keysize,
+                              NULL,
+                              0,
+                              (sgx_aes_gcm_data_ex_t *)cmk_blob);
+
+    memset_s(tmp, keysize, 0, keysize);
+
+    free(tmp);
+    return ret;
+}
+
+sgx_status_t ehsm_create_asymmetric_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_SUCCESS;
+
+    EVP_PKEY_CTX        *pkey_ctx        = NULL;
+    EVP_PKEY            *pkey           = NULL;
+    EC_GROUP            *ec_group       = NULL;
+    BIO                 *bio            = NULL;
+    uint8_t             *pem_keypair    = NULL;
+    uint32_t            key_len         = 0;
+
+    switch (cmk->metadata.keyspec) {
+        case EH_RSA_2048:
+        case EH_RSA_3072:
+        case EH_RSA_4096:
+            pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+            break;
+        case EH_EC_P224:
+        case EH_EC_P256:
+        case EH_EC_P384:
+        case EH_EC_P512:
+            pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+            break;
+        case EH_SM2:
+            pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_SM2, NULL);
+            break;
+        default:
+            break;
+    }
+    if (!pkey_ctx) {
+        goto out;
```

Comment:
the initial ret = SGX_SUCCESS, so you will return success when it met error?

Reply:
fixed